twenty-first/src/math/digest.rs

```diff
@@ -100,11 +101,18 @@ impl FromStr for Digest {
     type Err = TryFromDigestError;
 
     fn from_str(string: &str) -> Result<Self, Self::Err> {
-        let maybe_parsed_bfes: Result<Vec<_>, _> =
-            string.split(',').map(str::parse::<BFieldElement>).collect();
-        let parsed_bfes = maybe_parsed_bfes?;
-        let invalid_len_err = Self::Err::InvalidLength(parsed_bfes.len());
-        let digest_innards = parsed_bfes.try_into().map_err(|_| invalid_len_err)?;
+        let maybe_parsed_u64s: Result<Vec<_>, _> =
+            string.split(',').map(str::parse::<u64>).collect();
+        let parsed_u64s = maybe_parsed_u64s.map_err(ParseBFieldElementError::ParseU64Error)?;
+        let mut bfes = vec![];
+        for val in parsed_u64s {
+            if !BFieldElement::is_canonical(val) {
+                return Err(TryFromDigestError::NotCanonical(val));
+            }
+            bfes.push(BFieldElement::from(val))
+        }
```

Comment:
I'm a fan of iterator transforms. :blush: This is an opinion and not necessarily a change request.

```suggestion
        let bfe_try_from = |v: u64| -> Result<BFieldElement, _> {
            let bfe = BFieldElement::is_canonical(v).then(|| BFieldElement::new(v));
            bfe.ok_or(TryFromDigestError::NotCanonical(v))
        };
        let bfes: Vec<_> = parsed_u64s.into_iter().map(bfe_try_from).try_collect()?;
```

Upsides: :sunglasses:.
Downsides: not as early a return (I think).

Reply:
that's a clever transform.  I didn't immediately see an interator solution, hence the mutating for loop.

The tradeoff I see is that your solution is probably more rust-idiomatic and doesn't require a `mut`, while the for loop is more readable for people like myself coming from procedural languages.   I really have to study your transform to understand what it is doing, but it's good exercise for me, so I thank-you for making it.

I will use your version and add a comment to explain it's purpose.

early-return or not doesn't seem a concern with only 5 iterations.